### PR TITLE
feat: support customized kubelet configuration

### PR DESCRIPTION
**Pull request by mainred**
Besides customizing node configuration to change the default kubelet config, we introduce in this PR the ability to configure both the args of kubelet and download URL, as part of our custom configuration feature.

### CODE REVIEW

File: pkg/agent/utils.go [hunk 1]
```diff
@@ -358,18 +358,44 @@ func GetOrderedKubeletConfigFlagString(k map[string]string, cs *datamodel.Contai
 	kubeletConfigFileEnabled := IsKubeletConfigFileEnabled(cs, profile, kubeletConfigFileToggleEnabled)
 	keys := []string{}
 	for key := range k {
-		if !kubeletConfigFileEnabled || !TranslatedKubeletConfigFlags[key] {
-			keys = append(keys, key)
+		// TODO(qinhao): need to consider the case when enable-kubelet-config-file is enabled
+		if kubeletConfigFileEnabled && TranslatedKubeletConfigFlags[key] {
```
Comment (mainred) [inline pkg/agent/utils.go:362]: `&&`  is more straightforward.

Comment (mainred) [inline pkg/agent/utils.go:362]: Custom configuration includes not only the configuration of kubelet but also kube URL, as well as other components' images and configurations for internal use. These will be a conflict with [custom node configuration](https://docs.microsoft.com/en-us/azure/aks/custom-node-configuration), but the two features are not expected to be used for the same cluster.
TranslatedKubeletConfigFlags is introduced for dyanmic kubelet by https://github.com/Azure/AgentBaker/pull/139/files, and has been disabled by @anujmaheshwari1 in AKS

**Comment by mainred** [comment]
@alexeldeib @ganeshkumar5699 @feiskyer  please take a look.

File: pkg/agent/params.go [hunk 2]
```diff
@@ -62,6 +62,15 @@ func getParameters(config *datamodel.NodeBootstrappingConfiguration, generatorCo
 		}
 	}
 
+	if properties.CustomConfiguration != nil && properties.CustomConfiguration.KubernetesConfigurations != nil {
+		for componentName, configuration := range properties.CustomConfiguration.KubernetesConfigurations {
```
Comment (alexeldeib) [inline pkg/agent/params.go:66]: nit: why bother looping over a map when you are going to switch on component name anyway, and hardcode both componentName and customKubeBinaryUrl (i.e., don't use the keys for templating)

File: pkg/agent/utils.go [hunk 3]
```diff
@@ -370,6 +378,50 @@ func GetOrderedKubeletConfigFlagString(k map[string]string, cs *datamodel.Contai
 	return buf.String()
 }
 
+func getOrderedKubeletConfigFlagWithCustomConfigurationString(customConfig, defaultConfig map[string]string) string {
+	config := customConfig
+
+	if defaultConfig != nil {
+		for k, v := range defaultConfig {
+			// override configurations from default values to customized ones
+			if _, ok := config[k]; ok {
```
Comment (alexeldeib) [inline pkg/agent/utils.go:387]: nit: you don't need the if statement if you overwrite default config with custom config

on L395 you can range a nil map safely so it doesn't matter if defaultConfig is nil on L384 (assuming this is why you had it flipped to begin with)

**Review by alexeldeib — APPROVED** [review]
mostly nits, would appreciate an answer to https://github.com/Azure/AgentBaker/pull/1611#discussion_r838767904 though -- seems at a glance that extractKubeBinaries should work fine, not sure what the diff is here. but seems to work either way :)

File: parts/linux/cloud-init/artifacts/cse_install.sh [hunk 4]
```diff
@@ -227,21 +227,37 @@ extractHyperkube() {
...
+        # remove the kubelet binaries to make sure the only binary left is from the CUSTOM_KUBE_BINARY_DOWNLOAD_URL
+        rm -rf /usr/local/bin/kubelet-* /usr/local/bin/kubectl-*
+
+        # NOTE(mainred): we expect kubelet binary to be under `kubernetes/node/bin`. This suits the current setting of
+        # kube binaries used by AKS and Kubernetes upstream.
+        # TODO(mainred): let's see if necessary to auto-detect the path of kubelet
+        mkdir -p ${K8S_DOWNLOADS_DIR}
+        K8S_TGZ_TMP=${CUSTOM_KUBE_BINARY_DOWNLOAD_URL##*/}
+        retrycmd_get_tarball 120 5 "$K8S_DOWNLOADS_DIR/${K8S_TGZ_TMP}" ${CUSTOM_KUBE_BINARY_DOWNLOAD_URL} || exit $ERR_K8S_DOWNLOAD_TIMEOUT
+        tar -xzvf "$K8S_DOWNLOADS_DIR/${K8S_TGZ_TMP}" --strip-components=3 -C /usr/local/bin kubernetes/node/bin/kubelet kubernetes/node/bin/kubectl
```
[the remark this answers is not in the document]
Comment (mainred) [reply parts/linux/cloud-init/artifacts/cse_install.sh:242]: Thanks for pointing this out, will reuse `extractKubeBinaries`. My original thought is that we want to accept a tarball with a more flexible dir structure. But it seems at least for now it's not necessary as the tarball used by AKS and upstream are with the same structure.

File: pkg/agent/params.go (hunk 2 above)
Comment (mainred) [reply pkg/agent/params.go:66]: will use logic like `if val, ok := m["kubelet"]; ok` instead.

**Comment by mainred** [comment]
Thanks @alexeldeib. The change now looks more simple and straightfoward.

**Pull request merged**